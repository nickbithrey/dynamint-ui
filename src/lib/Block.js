import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Block = ({children, ...remainingProps}) => (
	<Stack.Item {...remainingProps}>
		{children}
	</Stack.Item>
);

export default Block;