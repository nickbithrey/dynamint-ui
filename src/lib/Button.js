import React from 'react';
import PropTypes from 'prop-types';
import { 
	PrimaryButton, 
	DefaultButton, 
	CommandBarButton, 
	IconButton 
} from 'office-ui-fabric-react/lib/Button';

const buttonTypes = {
	standard: DefaultButton,
	primary: PrimaryButton,
	command: CommandBarButton,
	icon: IconButton
};

const Button = ({text, onClick, type}) => {
	const Btn = buttonTypes[type];
	return <Btn text={text} onClick={onClick} />
};

Button.defaultProps = {
	type: 'standard'
};

Button.propTypes = {
	text: PropTypes.string.isRequired,
	onClick: PropTypes.func
}

export default Button;

export const newBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'standard'
	}
);

export const newPrimaryBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'primary'
	}
);