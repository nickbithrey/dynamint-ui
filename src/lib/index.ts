export interface Key {
    key: string | number;
}

export interface Condition<T> {
    condition: (input: T) => boolean;
}

export interface Loading {
    loading: boolean;
}

export interface Items<T> {
    items: Array<T>;
}

export interface URI {
    uri: string;
}

export interface Display {
    readOnly: boolean;
}

export interface IURILink {
    href: string;
}