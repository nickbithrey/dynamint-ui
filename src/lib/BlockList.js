import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'office-ui-fabric-react/lib/List';

const renderCell = (tag, updateEntry) => {
	const CellTag = tag;
	return (item, index) => {
		return <CellTag item={item} index={index} update={updateEntry} />;
	};
};

const BlockList = ({items, tag, updateEntry}) => {
	return (
		<List
	        items={items}
	        onRenderCell={renderCell(tag, updateEntry)}
	    />
	);
};

BlockList.propTypes = {
	items: PropTypes.arrayOf(PropTypes.object).isRequired,
	tag: PropTypes.func.isRequired
}

export default BlockList;