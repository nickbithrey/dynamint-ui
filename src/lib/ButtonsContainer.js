import React from 'react';
import Btn from '~/lib/Button';
import { Columnar, Block } from '~/lib/Grid';

const ButtonsContainer = ({children = [], buttons = []}) => (
	<Columnar horizontalAlign={'end'} tokens={{childrenGap: 5}}>
		{renderChildren(children)}
		{buttons.map(renderButton)}
	</Columnar>
);

const renderChildren = (children) => {
	if (Array.isArray(children)) {
		return children.map(renderChild)
	} else if (children) {
		return renderChild(children, 0);
	}
}

const renderChild = (child, i) => (
	<Block key={i}>
		{child}
	</Block>
);

const renderButton = (button, i) => {
	if (!button.condition()) {
		return;
	}
	return (
		<Block key={i}>
			<Btn {...button} />
		</Block>
	);
}

export default ButtonsContainer;