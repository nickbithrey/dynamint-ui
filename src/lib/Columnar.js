import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Columnar = ({children, ...remainingProps}) => (
	<Stack horizontal horizontalAlign="space-between" {...remainingProps}>
		{children}
	</Stack>
);

export default Columnar;