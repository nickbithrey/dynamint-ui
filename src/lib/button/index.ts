import { Button, IButton } from './Button';
import { ButtonContainer, IButtonContainer } from './ButtonContainer';

export { 
    Button,
    IButton,
    ButtonContainer,
    IButtonContainer
};