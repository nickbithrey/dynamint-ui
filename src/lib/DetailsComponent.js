import React from 'react';
import PropTypes from 'prop-types';
import { newBtn, newPrimaryBtn } from '~/lib/Button';
import Loading from '~/lib/Loading';
import { Rowise, Block } from '~/lib/Grid';
import ButtonsContainer from '~/lib/ButtonsContainer';

const detailsComponent = (WrappedComponent) => {
	class DetailsComponent extends React.Component {
		
		constructor(props) {
			super(props);
			this.load = this.load.bind(this);
			this.renderTitle = this.renderTitle.bind(this);
			this.createBackBtn = this.createBackBtn.bind(this);
			this.createSaveBtn = this.createSaveBtn.bind(this);
			this.back = this.back.bind(this);
			this.handleUpdateField = this.handleUpdateField.bind(this);
			this.handleUpdate = this.handleUpdate.bind(this);
			this.uri = props.location.state.uri;
			this.isCreate = props.location.state.isCreate;
		}
		
		load() {
			this.props.load(this.uri, this.isCreate);
		}
		
		componentDidMount() {
			this.load();
		}
		
		componentWillUnmount() {
			this.props.clear();
		}
		
		handleUpdateField(field) {
			this.props.updateField(field);
		}
		
		handleUpdate(e) {
			e.preventDefault();
			this.props.update(this.props.record, this.uri, this.isCreate)
			this.back();
		}
		
		createSaveBtn() {
			return newPrimaryBtn('Save', this.handleUpdate);
		}
		
		render() {
			if (this.props.loading || !this.props.record) {
				return <Loading />
			}
			const {
				updateField,
				update,
				...remainingProps
			} = this.props;
			return (
				<div>
					{this.renderTitle()}
					<Rowise tokens={{childrenGap: 5}}>
						<Block>
							<WrappedComponent fieldUpdate={this.handleUpdateField} update={this.handleUpdate} {...remainingProps} />
						</Block>
						<Block>
							<ButtonsContainer buttons={[this.createSaveBtn(), this.createBackBtn()]} />
						</Block>
					</Rowise>
				</div>
			);
		}
		
		renderTitle() {
			const {
				title
			} = this.props;
			return <h1>{title}</h1>;
		}
		
		back() {
			const {
				history
			} = this.props;
			history.goBack();
		}
		
		createBackBtn() {
			return newBtn('Back', this.back);
		}
		
	}

	DetailsComponent.propTypes = {
		load: PropTypes.func.isRequired,
		clear: PropTypes.func.isRequired,
		update: PropTypes.func.isRequired,
		updateField: PropTypes.func.isRequired,
		loading: PropTypes.bool.isRequired,
		history: PropTypes.object.isRequired,
		record: PropTypes.object,
		title: PropTypes.string.isRequired,
		url: PropTypes.string
	};
	
	return DetailsComponent;
}

export default detailsComponent;