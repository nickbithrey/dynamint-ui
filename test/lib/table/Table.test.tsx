import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import {
    Table,
    ITable,
    ITableColumn
} from 'lib/table/Table';
import { Selection } from 'office-ui-fabric-react/lib/DetailsList';

const expectedTag = 'StyledShimmeredDetailsListBase';

describe('<Table />', () => {
    
    const props: ITable = {
        loading: false,
        items: [
                {key: 'item1', col1: 'cell11', col2: 'cell12', col3: 'cell13'},
                {key: 'item2', col1: 'cell21', col2: 'cell22', col3: 'cell23'}
            ],
        columns: [
                  {name: 'column 1', key: 'col1', fieldName: 'col1', minWidth: 50},
                  {name: 'column 2', key: 'col2', fieldName: 'col2', minWidth: 60},
                  {name: 'column 3', key: 'col3', fieldName: 'col3', minWidth: 70}
            ],
        selection: new Selection()
    };

    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = shallow(<Table {...props} />);
    });
    
    it('renders a valid tag and matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    it('renders a valid tag and matches snapshot when no items are supplied', () => {
        wrapper.setProps({items: undefined});
        expect(wrapper.find(expectedTag).prop('items')).toHaveLength(0);
    });
    
});