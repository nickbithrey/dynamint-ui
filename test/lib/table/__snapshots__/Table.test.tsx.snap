// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> renders a valid tag and matches snapshot 1`] = `
<StyledShimmeredDetailsListBase
  columns={
    Array [
      Object {
        "fieldName": "col1",
        "key": "col1",
        "minWidth": 50,
        "name": "column 1",
      },
      Object {
        "fieldName": "col2",
        "key": "col2",
        "minWidth": 60,
        "name": "column 2",
      },
      Object {
        "fieldName": "col3",
        "key": "col3",
        "minWidth": 70,
        "name": "column 3",
      },
    ]
  }
  enableShimmer={false}
  items={
    Array [
      Object {
        "col1": "cell11",
        "col2": "cell12",
        "col3": "cell13",
        "key": "item1",
      },
      Object {
        "col1": "cell21",
        "col2": "cell22",
        "col3": "cell23",
        "key": "item2",
      },
    ]
  }
  selection={
    Selection {
      "_anchoredIndex": 0,
      "_canSelectItem": [Function],
      "_changeEventSuppressionCount": 0,
      "_exemptedCount": 0,
      "_exemptedIndices": Object {},
      "_getKey": [Function],
      "_isModal": false,
      "_items": Array [],
      "_keyToIndexMap": Object {},
      "_onSelectionChanged": undefined,
      "_selectedItems": null,
      "_unselectableCount": 0,
      "_unselectableIndices": Object {},
      "count": 0,
      "mode": 2,
    }
  }
  selectionPreservedOnEmptyClick={true}
/>
`;
